File: solutions/go/protein-translation/2/protein_translation.go
// Translate RNA sequences into proteins.
package protein

import "errors"

var (
	ErrInvalidBase = errors.New("ErrInvalidBase")
	ErrStop        = errors.New("ErrStop")
)

// FromRNA Translate RNA sequences into Codon sequences and this into Proteins
func FromRNA(rna string) ([]string, error) {
	var proteins []string
	for i := 0; i < len(rna); i += 3 {

		if protein, err := FromCodon(rna[i:(i + 3)]); err == nil {
			proteins = append(proteins, protein)
		} else if err == ErrStop {
			return proteins, nil
		} else {
			return proteins, err
		}
	}
	return proteins, nil
}

// FromCodon Translate Codon sequences into proteins.
func FromCodon(codon string) (string, error) {
	switch codon {
	case "AUG":
		return "Methionine", nil
	case "UUU", "UUC":
		return "Phenylalanine", nil
	case "UUA", "UUG":
		return "Leucine", nil
	case "UCU", "UCC", "UCA", "UCG":
		return "Serine", nil
	case "UAU", "UAC":
		return "Tyrosine", nil
	case "UGU", "UGC":
		return "Cysteine", nil
	case "UGG":
		return "Tryptophan", nil
	case "UAA", "UAG", "UGA":
		return "STOP", ErrStop
	default:
		return "", ErrInvalidBase
	}
}

File: solutions/csharp/protein-translation/2/ProteinTranslation.cs
public static class ProteinTranslation
{
    public static string[] Proteins(string strand)
    {
        var proteins = new List<string>();
        for (int i = 0; i < strand.Length; i += 3)
        {
            var codon = strand.Substring(i, 3);
            if (CodonToProtein().TryGetValue(codon, out var protein))
                proteins.Add(protein);
            else if (codon == "UAA" || codon == "UAG" || codon == "UGA")
                break;
        }
        return [.. proteins];
    }

    private static Dictionary<string, string> CodonToProtein() => new()
        {
            { "AUG", "Methionine"},
            { "UUU", "Phenylalanine"},
            { "UUC", "Phenylalanine"},
            { "UUA", "Leucine"},
            { "UUG", "Leucine"},
            { "UCU", "Serine"},
            { "UCC", "Serine"},
            { "UCA", "Serine"},
            { "UCG", "Serine" },
            { "UAU", "Tyrosine"},
            { "UAC", "Tyrosine" },
            { "UGU", "Cysteine"},
            { "UGC", "Cysteine" },
            { "UGG", "Tryptophan" }
        };
}